Api/news list polymorphism

timotiusnc opened the pull request:
I changed the structure of polymorphism a bit (refer to #8)

Closes #8

timotiusnc committed:
1. Add Item base class
2. Modify News class to extend Item
3. Update the getter and setter of News class

timotiusnc committed:
1. Add class News, Comment, Job, Poll, and PollOpt.
2. Add enum class Type as a public inner class of class Item.
3. Add factory method `valueOf` in class Item to instantiate Item based on its type.

timotiusnc committed:
1. Update the polymorphism structure (refer to issue #8)
2. Retrieve some news online and display it on NewsListFragment

wiseodd commented on app/src/main/java/io/pumpkinz/pumpkinreader/model/Comment.java:
```diff
@@ -1,37 +1,55 @@
+    int parent;
+    List<Integer> kids;
+    Comment parentComment;
+    List<Comment> comments;
```
Why does a comment consist of N comments?

timotiusnc replied:
Hacker News' comment does have N comments. [See here](https://hacker-news.firebaseio.com/v0/item/2921983.json?print=pretty).

wiseodd replied:
That's the kids right?
Not the comment itself.

timotiusnc replied:
Yes, but eventually we will need to retrieve and store it in an object.
Or do you think we should store it somewhere else?

wiseodd replied:
The problem is the naming. 
It's not descriptive enough.

timotiusnc replied:
OK, I'll change it

wiseodd replied:
See my note here https://github.com/pumpkinz/pumpkin-reader/pull/18#discussion_r33845944

wiseodd commented on app/src/main/java/io/pumpkinz/pumpkinreader/model/Item.java:
```diff
@@ -0,0 +1,129 @@
+import java.util.Date;
+
+
+public abstract class Item {
```
Why do you have Item and ItemPOJO?

timotiusnc replied:
ItemPOJO is a class to contain the return value of Retrofit API. Retrofit will automatically parse the JSON, instantiate the class and map its JSON values into the class' fields (in this case, the ItemPOJO class).

Whereas Item is the real class to represent a Hacker News Item.

wiseodd replied:
Couldn't we just use the ItemPOJO for the base class, or Item as Retrofit container?
I think this violates DRY.

timotiusnc replied:
We can't use Item as a return value for Retrofit container because we will only get `id, deleted, type, by, time, text, dead` (Item's fields) and lose the others (`url, title`, etc).

We can't use its child class (e.g. `Story, Job, etc`) either as we don't know the type yet.

timotiusnc replied:
Hemm, but maybe I can tap the Retrofit return value and instantiate the object manually. I'll take a look.

wiseodd replied:
Or maybe Item could inherit ItemPOJO?

Also BTW, what is POJO naming best practice?
I think it's the same with model naming convention, as they are POJO alike.
But CMIIW.

wiseodd commented on app/src/main/java/io/pumpkinz/pumpkinreader/NewsListFragment.java:
```diff
@@ -98,9 +103,10 @@ public void onError(Throwable e) {
+            public void onNext(List<News> items) {
+                Log.d("stories", String.valueOf(items.size()));
+                Log.d("stories content", items.toString());
+                dataset.addAll(items);
```
I prefer NewsAdapter to expose a public method to add a new items to dataset, instead of we rely on side effect like this.

timotiusnc replied:
All right

timotiusnc replied:
Move observable to NewsAdapter alltogether.
NewsListFragment doesn't need to know about List<News> dataset at all.

wiseodd commented on app/src/main/java/io/pumpkinz/pumpkinreader/model/News.java:
```diff
@@ -1,62 +1,90 @@
-    Date date;
     int score;
+    String title;
+    int descendants;
```
This naming is also misleading.
If I think of an object's descendants, then I will think of another objects.
Not how many descendants the object has.

wiseodd replied:
Ah I see why the naming is so no-no.
You're following the JSON field.

I think we have to separate them.
They have difference concern, and we shouldn't make it coupled.

timotiusnc replied:
Difference concern? What difference concern?

Or do you mean `different` concern?

wiseodd replied:
Yes different, sorry.

timotiusnc replied:
By the way, we can just remove `descendants` as we can always do `kids.size()`.

What do you think?

wiseodd replied:
Yes, that's actually the best practice.

`Never create an acessor that could be computed from existing acessor` 
or something along that line.

timotiusnc committed:
1. Remove ItemPOJO
2. Use custom Gson Deserializer for Item and News
3. Update all model class (remove constructor with ItemPOJO)

timotiusnc committed:
Add public method in NewsAdapter to add some News into its dataset

timotiusnc commented:
I ended up only exposing public method on NewsAdapter.
It's not possible to move the subscription to NewsAdapter as it is coupled into the Fragment view: we need to bind the subscription to the fragment, do unsubscription when the view is destroyed, etc.

Ready to be reviewed!

timotiusnc committed:
Filter out NULL news from getTopNews method

timotiusnc committed:
Remove `descendants` attribute from News class
